src/components/Table/GridTable.tsx

```diff
@@ -359,16 +365,38 @@ export function GridTable<R extends Kinded, S = {}, X extends Only<GridTableXss,
     return rows;
   }, [columns, rows, sorting, sortState]);
 
-  const columnSizes = useSetupColumnSizes(style, columns, tableRef, resizeTarget);
+  // Filter rows - ensures parent rows remain in the list if any children match the filter.
+  const filterRows: (acc: ParentChildrenTuple<R>[], row: GridDataRow<R>) => ParentChildrenTuple<R>[] = useCallback(
+    (acc: ParentChildrenTuple<R>[], row: GridDataRow<R>) => {
+      // Break up "foo bar" into `[foo, bar]` and a row must match both `foo` and `bar`
+      const filters = (filter && filter.split(/ +/)) || [];
+      const matches =
+        row.kind === "header" ||
+        filters.length === 0 ||
+        !!row.pin ||
+        filters.every((f) =>
+          columns.map((c) => applyRowFn(c, row, api)).some((maybeContent) => matchesFilter(maybeContent, f)),
+        );
 
-  // Make a single copy of our current collapsed state, so we'll have a single observer.
-  const collapsedIds = useComputed(() => rowState.collapsedIds, [rowState]);
+      if (matches) {
+        return acc.concat([[row, row.children?.reduce(filterRows, []) ?? []]]);
+      }
 
-  // Filter + flatten + component-ize the sorted rows.
-  let [headerRows, filteredRows]: [RowTuple<R>[], RowTuple<R>[]] = useMemo(() => {
-    // Break up "foo bar" into `[foo, bar]` and a row must match both `foo` and `bar`
-    const filters = (filter && filter.split(/ +/)) || [];
+      const isCollapsed = collapsedIds.includes(row.id);
```

Comment:
Maybe put this chunk into an `else` of the `if matches`? I just initially was getting concerned we were adding `row` to `acc` twice, b/c we do it on line 382 as well as 389, but was missing "oh right", line 382 is an early return.